### PR TITLE
#9 Add Assert.Equal migration with comparer

### LINKED ISSUE #9: Add Assert.Equal migration with comparer
```csharp
Assert.Equal(expected, actual, _comparer);
```

Can be replaced with next code:

```csharp
actual.Should().BeEquivalentTo(expected, x => x.Using(_comparer));
```

### PR DESCRIPTION
**Pull request by inyutin-maxim**
Fix #9

### CODE REVIEW

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs [hunk 1]
```diff
@@ -27,10 +28,10 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
         if (nodeToFix == null)
             return;
 
-        var title = "Use FluentAssertions";
+        const string title = "Use FluentAssertions";
         var codeAction = CodeAction.Create(
             title,
-            ct => Rewrite(context.Document, nodeToFix, context.CancellationToken),
+            _ => Rewrite(context.Document, nodeToFix, context.CancellationToken),
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]:
```suggestion
            cancellationToken => Rewrite(context.Document, nodeToFix, cancellationToken),
```

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs [hunk 2]
```diff
@@ -152,6 +153,17 @@ private static async Task<Document> Rewrite(Document document, SyntaxNode nodeTo
                 MemberAccessExpression(InvokeShould(originalArguments[1]), newMethodName),
                 ArgumentList(originalMethod.ArgumentList.Arguments[0].Expression, originalMethod.ArgumentList.Arguments[2].Expression));
         }
+        else if (methodName is "Equal" or "NotEqual" && method.Parameters.Length == 3 && method.Parameters[2].Type.Name == nameof(IEqualityComparer))
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]: You must compare symbols correctly to avoid false-positive

```suggestion
        else if (methodName is "Equal" or "NotEqual" && method.Parameters.Length == 3 && method.Parameters[2].Type.Equals(compilation, "System.Collections.Generic.IEqualityComparer`1"))
```

File: Meziantou.FluentAssertionsAnalyzers.Tests/XUnitToFluentAssertionsAnalyzerUnitTests.cs [hunk 3]
```diff
@@ -510,6 +510,123 @@ public void MyTest()
...
+			return x.Name == y.Name;
+		}
+
+		public int GetHashCode(TestObject obj)
+		{
+			return obj.Name != null
+				? obj.Name.GetHashCode()
+				: 0;
+		}
+	}
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers.Tests/XUnitToFluentAssertionsAnalyzerUnitTests.cs]:
```suggestion
		public bool Equals(TestObject x, TestObject y)
The implementation of this object doesn't matter. You can remove it.

```suggestion
		public bool Equals(TestObject x, TestObject y) => throw null;
```

**Review by meziantou — CHANGES_REQUESTED** [review]
Thanks for your contribution! Some minor changes before I can merge the PR

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs [hunk 4]
```diff
@@ -27,10 +28,10 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
         if (nodeToFix == null)
             return;
 
-        var title = "Use FluentAssertions";
+        const string title = "Use FluentAssertions";
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]:
```suggestion
        const string Title = "Use FluentAssertions";
```

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs (hunk 2 above)
Comment (inyutin-maxim) [reply Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]: hm, this suggestion does not work. I thinking about it

Comment (meziantou) [reply Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]: There is a similar check above, maybe it can give you an example. If it doesn't work, I can have a look at it later

https://github.com/meziantou/Meziantou.FluentAssertionsAnalyzers/pull/10/files/99e88cb5a7fba0f94013f8b96cbbf8f30ae75171#diff-82ee5eb7e03488eb286be3ef4e07f18a6728aab925f72ca5c64a886702bf4cffR101

Comment (inyutin-maxim) [reply Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]: This worked with code:
```csharp
 else if (methodName is "Equal" or "NotEqual" && method.Parameters.Length == 3 && method.Parameters[2].Type.OriginalDefinition.IsOrImplements(compilation, "System.Collections.Generic.IEqualityComparer`1"))
```

**Comment by inyutin-maxim** [comment]
Thanks for review, i think now it better))

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs (hunk 2 above)
Comment (meziantou) [reply Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]: `Assert.Equal` is a generic method, so you need to get the original definition of the symbol to convert `IEqualityComparer<SpecificType>` to `IEqualityComparer<T>`. Then the comparison should work.

````
method.Parameters[2].Type.OriginalDefinition.Equals(...);
````

File: Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs [hunk 5]
```diff
@@ -27,11 +27,11 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
             return;
 
-        var title = "Use FluentAssertions";
+        const string Title = "Use FluentAssertions";
         var codeAction = CodeAction.Create(
-            title,
-            ct => Rewrite(context.Document, nodeToFix, context.CancellationToken),
-            equivalenceKey: title);
+            Title,
+            cancellationToken  => Rewrite(context.Document, nodeToFix, cancellationToken ),
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers/XUnitAssertAnalyzerCodeFixProvider.cs]:
```suggestion
            cancellationToken => Rewrite(context.Document, nodeToFix, cancellationToken),
```

File: Meziantou.FluentAssertionsAnalyzers/Meziantou.FluentAssertionsAnalyzers.csproj [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (meziantou) [inline Meziantou.FluentAssertionsAnalyzers/Meziantou.FluentAssertionsAnalyzers.csproj]: Revert this change to fix the build

Comment (inyutin-maxim) [reply Meziantou.FluentAssertionsAnalyzers/Meziantou.FluentAssertionsAnalyzers.csproj]: I'm sorry, I thought that since the project has only one target, it's redundant.
